src/BinFile/WasmParser.fs

```diff
@@ -0,0 +1,367 @@
+(*
+  B2R2 - the Next-Generation Reversing Platform
+
+  Copyright (c) SoftSec Lab. @ KAIST, since 2016
+
+  Permission is hereby granted, free of charge, to any person obtaining a copy
+  of this software and associated documentation files (the "Software"), to deal
+  in the Software without restriction, including without limitation the rights
+  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+  copies of the Software, and to permit persons to whom the Software is
+  furnished to do so, subject to the following conditions:
+
+  The above copyright notice and this permission notice shall be included in all
+  copies or substantial portions of the Software.
+
+  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+  SOFTWARE.
+*)
+
+module internal B2R2.BinFile.Wasm.Parser
+
+open B2R2
+open B2R2.BinFile
+open B2R2.BinFile.Wasm.Section
+open System
+
+let sectionIdToName (secId: SectionId) (off: int) =
+  match secId with
+  | SectionId.Custom -> String.Format ("custom_{0:x}", off)
+  | SectionId.Type -> "type"
+  | SectionId.Import -> "import"
+  | SectionId.Function -> "function"
+  | SectionId.Table -> "table"
+  | SectionId.Memory -> "memory"
+  | SectionId.Global -> "global"
+  | SectionId.Export -> "export"
+  | SectionId.Start -> "start"
+  | SectionId.Element -> "element"
+  | SectionId.Code -> "code"
+  | SectionId.Data -> "data"
+  | _ -> ""
+
+let private summerizeSections (reader: BinReader) offset =
+  let rec loop (acc: _ list) no =
+    if reader.IsOutOfRange no then
+      acc
+    else
+      let id, size, len = peekSectionHeader reader no
+      let headerSize = len + 1
+      let no' = no + headerSize + int size
+      let summary = {
+        Id = id
+        Name = sectionIdToName id no
+        Offset = no
+        HeaderSize = uint32 headerSize
+        ContentsSize = size
+      }
+      loop (acc @ [summary]) no'
```

Comment:
Please use consing (`::`) and then `List.rev` later when the function returns. Appending (`@`) in each loop is much less efficient than consing + rev.